[WPE] Implement SystemSettingsManagerProxyWPE

TingPing opened the pull request:
#### 69b48be69d3d8a3e541c45cdf5b5d4c72dc1fc57
<pre>
[WPE] Implement SystemSettingsManagerProxyWPE
<a href="https://bugs.webkit.org/show_bug.cgi?id=282517">https://bugs.webkit.org/show_bug.cgi?id=282517</a>

Reviewed by Carlos Garcia Campos..

This implements the various settings for WPEPlatform.

The types and names are not exact matches for GTK as
they could be simplified or turned into enums which
is more friendly. Internally they are converted to the
same format as GTK.

* Source/WebCore/platform/glib/SystemSettings.cpp:
(WebCore::SystemSettings::defaultSystemFont const):
* Source/WebCore/platform/glib/SystemSettings.h:
* Source/WebCore/platform/graphics/freetype/FontCacheFreeType.cpp:
(WebCore::getFamilyNameStringFromFamily):
* Source/WebCore/platform/graphics/skia/FontCacheSkia.cpp:
(WebCore::getFamilyNameStringFromFamily):
* Source/WebCore/platform/gtk/GtkUtilities.cpp:
(WebCore::defaultGtkSystemFont): Deleted.
* Source/WebCore/platform/gtk/GtkUtilities.h:
* Source/WebKit/SourcesWPE.txt:
* Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
(webkitWebViewConstructed):
* Source/WebKit/UIProcess/glib/SystemSettingsManagerProxy.cpp:
* Source/WebKit/UIProcess/glib/SystemSettingsManagerProxy.h:
* Source/WebKit/UIProcess/wpe/SystemSettingsManagerProxyWPE.cpp: Added.
(WebKit::SystemSettingsManagerProxy::themeName const):
(WebKit::SystemSettingsManagerProxy::darkMode const):
(WebKit::SystemSettingsManagerProxy::fontName const):
(WebKit::SystemSettingsManagerProxy::xftAntialias const):
(WebKit::SystemSettingsManagerProxy::xftHinting const):
(WebKit::SystemSettingsManagerProxy::xftHintStyle const):
(WebKit::SystemSettingsManagerProxy::xftRGBA const):
(WebKit::SystemSettingsManagerProxy::xftDPI const):
(WebKit::SystemSettingsManagerProxy::followFontSystemSettings const):
(WebKit::SystemSettingsManagerProxy::cursorBlink const):
(WebKit::SystemSettingsManagerProxy::cursorBlinkTime const):
(WebKit::SystemSettingsManagerProxy::primaryButtonWarpsSlider const):
(WebKit::SystemSettingsManagerProxy::overlayScrolling const):
(WebKit::SystemSettingsManagerProxy::enableAnimations const):
(WebKit::SystemSettingsManagerProxy::SystemSettingsManagerProxy):
* Source/WebKit/WPEPlatform/wpe/WPESettings.cpp:
* Tools/MiniBrowser/wpe/main.cpp:
(loadConfigFile):
(activate):

Canonical link: <a href="https://commits.webkit.org/286738@main">https://commits.webkit.org/286738@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/01725d3b1133fc41a56081405e28e023d0d46585

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/76876 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/55911 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/29782 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/81418 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/28152 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/78993 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/65053 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/4204 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/60237 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/18318 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/79943 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/50185 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/65993 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/40554 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/47587 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/23488 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/26489 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/68704 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/23817 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/82856 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/4252 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/2825 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/68514 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/4406 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/65966 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/67769 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16900 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/11753 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/9837 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/4199 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/7012 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/4222 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/7653 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/5980 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

carlosgcampos commented on Source/WebKit/UIProcess/wpe/SystemSettingsManagerProxyWPE.cpp:
```diff
@@ -0,0 +1,146 @@
+#include <wpe/WPESettings.h>
+#include <wtf/glib/GUniquePtr.h>
+
+
```
```suggestion
```

carlosgcampos commented on Source/WebKit/UIProcess/wpe/SystemSettingsManagerProxyWPE.cpp:
```diff
@@ -0,0 +1,146 @@
+
+#define SETTING_FONT_NAME "/wpe-platform/font-name"
+#define SETTING_DARK_MODE "/wpe-platform/dark-mode"
+#define SETTING_DISABLE_ANIMATIONS "/wpe-platform/disable-animations"
```
For well-known settings registered by WPE platform, the idea is to make these public and documented.

carlosgcampos commented on Source/WebKit/UIProcess/wpe/SystemSettingsManagerProxyWPE.cpp:
```diff
@@ -0,0 +1,146 @@
+        GUniqueOutPtr<GError> error;
+        if (!wpe_settings_register(m_settings, setting.key, setting.type, setting.defaultValue, &error.outPtr()))
+            g_warning("Failed to register setting %s: %s", setting.key, error->message);
+    }
```
I think these should be registered by wpe platform, not by WebKit. Because they are read by WebKit, but they are supposed to be written by the platforms. For example, a GTK4 platform would write them from GtkSettings.

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
```diff
@@ -905,6 +906,8 @@ static void webkitWebViewConstructed(GObject* object)
         }
     }
+
+    SystemSettingsManagerProxy::initialize();
```
This should only be called when the new api is used, you should check if priv->display is not null

carlosgcampos commented on Source/WebKit/UIProcess/wpe/SystemSettingsManagerProxyWPE.cpp:
```diff
@@ -0,0 +1,208 @@
+#include <wtf/glib/GUniquePtr.h>
+
+static String
+getString(WPESettings* settings, const char* key, const String& defaultValue)
```
```suggestion
static String getString(WPESettings* settings, const char* key, const String& defaultValue)
```

carlosgcampos commented on Source/WebKit/UIProcess/wpe/SystemSettingsManagerProxyWPE.cpp:
```diff
@@ -0,0 +1,208 @@
+}
+
+static bool
+getBool(WPESettings* settings, const char* key, bool defaultValue)
```
```suggestion
static bool getBool(WPESettings* settings, const char* key, bool defaultValue)
```

carlosgcampos commented on Source/WebKit/UIProcess/wpe/SystemSettingsManagerProxyWPE.cpp:
```diff
@@ -0,0 +1,208 @@
+}
+
+static uint32_t
+getUint32(WPESettings* settings, const char* key, uint32_t defaultValue)
```
```suggestion
static uint32_t getUint32(WPESettings* settings, const char* key, uint32_t defaultValue)
```

carlosgcampos commented on Source/WebKit/UIProcess/wpe/SystemSettingsManagerProxyWPE.cpp:
```diff
@@ -0,0 +1,208 @@
+}
+
+static uint8_t
+getUint8(WPESettings* settings, const char* key, uint8_t defaultValue)
```
```suggestion
static uint8_t getUint8(WPESettings* settings, const char* key, uint8_t defaultValue)
```

carlosgcampos commented on Source/WebKit/UIProcess/wpe/SystemSettingsManagerProxyWPE.cpp:
```diff
@@ -0,0 +1,208 @@
+}
+
+static double
+getDouble(WPESettings* settings, const char* key, double defaultValue)
```
```suggestion
static double getDouble(WPESettings* settings, const char* key, double defaultValue)
```

carlosgcampos commented on Source/WebKit/UIProcess/wpe/SystemSettingsManagerProxyWPE.cpp:
```diff
@@ -0,0 +1,208 @@
+    return value;
+}
+
+namespace WebKit {
```
Why aren't previous helpers in WebKit namespace?

carlosgcampos commented on Source/WebKit/UIProcess/wpe/SystemSettingsManagerProxyWPE.cpp:
```diff
@@ -0,0 +1,208 @@
+}
+
+SystemSettingsManagerProxy::SystemSettingsManagerProxy()
+    : m_settings(wpe_display_get_settings(wpe_display_get_default()))
```
This is not correct, if the view is created with a non-default display, we would create here a new display only used to get its settings. I don't know what to do because settings are supposed to be global. For now we can just use the primary display.

TingPing replied:
Yes the GTK port has the same problem (and always didn't follow the current GdkDisplay).

carlosgcampos commented on Source/WebKit/UIProcess/wpe/SystemSettingsManagerProxyWPE.cpp:
```diff
@@ -0,0 +1,208 @@
+SystemSettingsManagerProxy::SystemSettingsManagerProxy()
+    : m_settings(wpe_display_get_settings(wpe_display_get_default()))
+{
+    ASSERT(m_settings);
```
This is true if SystemSettingsManagerProxy is only used when the new api is used.

TingPing committed:
[WPE] Implement SystemSettingsManagerProxyWPE
https://bugs.webkit.org/show_bug.cgi?id=282517

Reviewed by Carlos Garcia Campos..

This implements the various settings for WPEPlatform.

The types and names are not exact matches for GTK as
they could be simplified or turned into enums which
is more friendly. Internally they are converted to the
same format as GTK.

* Source/WebCore/platform/glib/SystemSettings.cpp:
(WebCore::SystemSettings::defaultSystemFont const):
* Source/WebCore/platform/glib/SystemSettings.h:
* Source/WebCore/platform/graphics/freetype/FontCacheFreeType.cpp:
(WebCore::getFamilyNameStringFromFamily):
* Source/WebCore/platform/graphics/skia/FontCacheSkia.cpp:
(WebCore::getFamilyNameStringFromFamily):
* Source/WebCore/platform/gtk/GtkUtilities.cpp:
(WebCore::defaultGtkSystemFont): Deleted.
* Source/WebCore/platform/gtk/GtkUtilities.h:
* Source/WebKit/SourcesWPE.txt:
* Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
(webkitWebViewConstructed):
* Source/WebKit/UIProcess/glib/SystemSettingsManagerProxy.cpp:
* Source/WebKit/UIProcess/glib/SystemSettingsManagerProxy.h:
* Source/WebKit/UIProcess/wpe/SystemSettingsManagerProxyWPE.cpp: Added.
(WebKit::SystemSettingsManagerProxy::themeName const):
(WebKit::SystemSettingsManagerProxy::darkMode const):
(WebKit::SystemSettingsManagerProxy::fontName const):
(WebKit::SystemSettingsManagerProxy::xftAntialias const):
(WebKit::SystemSettingsManagerProxy::xftHinting const):
(WebKit::SystemSettingsManagerProxy::xftHintStyle const):
(WebKit::SystemSettingsManagerProxy::xftRGBA const):
(WebKit::SystemSettingsManagerProxy::xftDPI const):
(WebKit::SystemSettingsManagerProxy::followFontSystemSettings const):
(WebKit::SystemSettingsManagerProxy::cursorBlink const):
(WebKit::SystemSettingsManagerProxy::cursorBlinkTime const):
(WebKit::SystemSettingsManagerProxy::primaryButtonWarpsSlider const):
(WebKit::SystemSettingsManagerProxy::overlayScrolling const):
(WebKit::SystemSettingsManagerProxy::enableAnimations const):
(WebKit::SystemSettingsManagerProxy::SystemSettingsManagerProxy):
* Source/WebKit/WPEPlatform/wpe/WPESettings.cpp:
* Tools/MiniBrowser/wpe/main.cpp:
(loadConfigFile):
(activate):

Canonical link: https://commits.webkit.org/286738@main
